### ISSUE TITLE
Having OBS installed when developing Vulkan apps messes with the Vulkan validation layer

### ISSUE THREAD

**Issue by RikoOphorst** [obsproject/obs-studio#3145]
Sorry, I'm not comforming to your bug report procedures, since this is relating to system stability when having OBS installed on the machine (and not even running!!!!!!!!!!).

For the past couple months, we've seen very odd crashes caused by Vulkan's validation layer. Specifically, it was telling us this:

`VUID-vkGetSwapchainImagesKHR-commonparent(ERROR / SPEC): msgNum: 1944321689 - Validation Error: [ VUID-vkGetSwapchainImagesKHR-commonparent ] Object 0: handle = 0x15f4cab3db0, type = VK_OBJECT_TYPE_INSTANCE; | MessageID = 0x73e3fe99 | Object 0x15f61c6c578 of type VkSwapchainKHR was not created, allocated or retrieved from the correct device. The Vulkan spec states: Both of device, and swapchain must have been created, allocated, or retrieved from the same VkInstance (https://vulkan.lunarg.com/doc/view/1.2.141.0/windows/1.2-extensions/vkspec.html#VUID-vkGetSwapchainImagesKHR-commonparent)`

which is very odd for a couple reasons:

- we only had 1 single VkInstance, we never made any other one - though the validation layer was complaining that we were creating a swapchain from a device that was not created on the same `VkInstance`?
- we do have multiple `VkDevice` objects that we created, but we absolutely 100% made sure that we created the swapchain from the correct device, from the correct command queue.

but ABSOLUTELY the weirdest part of this bug:

the error occurred from `vkCreateSwapchainKHR` - the _creation_ of the swapchain.
however, the validation layer error was telling us that the error was coming from `vkGetSwapchainImagesKHR` - the _retrieval_ of swapchain images.

???

For months we were clueless as to why this was possibly happening. It wasn't consistently happening, but sometimes it did.

After some time, we decided to delve a little deeper into this issue...
and then we noticed this, upon our swapchain creation: 

![image](https://user-images.githubusercontent.com/6514077/87052163-c343a400-c200-11ea-8cc1-d856ae9323da.png)

somehow, from vulkan-1.dll, we ended up in another DLL, "graphics-hook64.dll" - which suspiciously isn't part of the Vulkan eco-system, but from the OBS software installed on our machines.

Upon deleting the OBS software from the machine, the error no longer happened! The validation layer no longer complained, everything was happy.

So our thought here is that your graphics-hook64.dll somehow isn't dealing well with Vulkan apps having complicated VkInstance, VkPhysicalDevice, VkDevice and VkSwapChainKHR configurations.

I am not familiar with the source code behind OBS, but this is something that should be fixed as soon as possible. Many vulkan developers are possibly running into this stuff if they have OBS installed on their machines.

**Comment by dodgepong** [comment]
Why would an issue "relating to system stability when having OBS installed" preclude you from filling out the bug report template?

**Comment by RikoOphorst** [comment]
@dodgepong Is that really a problem? I'm not familiar with this entire project except that I occasionally use it to record some videos. 
Finding out that this piece of software, seemingly completely unrelated to what I was doing somewhere else on my machine, was somehow messing with what I was doing.

I feel like the description given in the original post gives you all the info that I _can_ give you. We spent **months** dealing with random crashes, caused by some odd-ball piece of software messing with another software install on the machine.

If it's really that big a problem, I can add the few bug report template headers, but they functionally/effectively won't add any other info that you couldn't already derive from this post. Given the NDA restrictions I am under, I cannot give you more info what I already did.

Personally, I would take a look at the problem at hand, see what you guys can do on your side, and not worry too much about a report on a serious stability issue that is not using the proper bug reporting template.

**Comment by RytoEX** [comment]
> I feel like the description given in the original post gives you all the info that I _can_ give you.

What version of Windows was this?  Windows 7, 8, 8.1, 10?  What version of OBS did you have installed?  Was it a release version or a self-compiled version?

Have you tested the latest OBS git source code to see if we have already fixed this?  There have been [multiple updates to the Vulkan-related code](https://github.com/obsproject/obs-studio/pulls?q=is%3Apr+is%3Aclosed+created%3A%3E2020-03-17+vulkan) since the latest OBS release.

Have you checked if any pending pull requests have solved this problem?  There are [a few unmerged PRs for Vulkan-related code](https://github.com/obsproject/obs-studio/pulls?q=is%3Apr+is%3Aopen+created%3A%3E2020-03-17+vulkan).

> Personally, I would take a look at the problem at hand, see what you guys can do on your side, and not worry too much about a report on a serious stability issue that is not using the proper bug reporting template.

We have a template to make it easier to pick out the important pieces of information from a bug report.  Having the template filled out helps us review the information more quickly and therefore helps us investigate the reported issue.

Lastly, this is a volunteer-driven open source software project.  There are perhaps two people involved with the project who have a good understanding of the Vulkan-related code.  They may look into this when their time allows them to do so.

**Comment by dodgepong** [comment]
At the very least, it's a matter of mutual respect. By submitting a bug report, you're asking someone to volunteer to spend time investigating an issue and as follow-up questions. But when I see someone delete the issue template and instead just post whatever they want, the message I get from that is that they don't want to spend the time to fill out the template because they deem their time more valuable than ours.

The issue template is designed to get the most important information that we need to investigate the issue. For example, your report lacks the version of OBS you had installed (maybe you're using an outdated version? or maybe you built it from source?) and it doesn't have clear reproduction steps for the issue (it's sort of a pile of rambling prose). That's a full half of the issue template that we highly value, and now we have to waste time asking for that information again when we had already asked for it in the first place via the issue template.

I understand that your NDA may prevent you from revealing some contextual information about the issue. However, without knowing details about how you're using Vulkan (i.e. "complicated VkInstance, VkPhysicalDevice, VkDevice and VkSwapChainKHR configurations"), for all I know it could be difficult to pin down the cause. Because of that, the reproduction steps are pretty important here, and even a "toy" example that reproduces the issue would be valuable. Or maybe it's an obvious thing, I don't know. I'll leave that to @jpark37 to decide, though, as he is the Vulkan expert on the team.

tl;dr The best way to make sure that we are willing to spend the time to investigate an issue is for you to do your own due diligence and spend the time to properly fill out the issue template.

**Comment by jpark37** [comment]
Skimming back through the data structures, it looks like we aren't well equipped to deal with more than 16 devices or swap chains. If you have more than that, then I see what the issue is, and we can fix up the hook to dynamically allocate instead.

What we do isn't that different from the way Steam overlay hooks into a game process for example. We just want to extract the swap chain image rather than write to it.

The hook is designed for low impact capture of games. For D3D/GL, OBS will inject on the fly, but Vulkan provides an official framework for hooking API calls. We use what they call an "implicit layer," which means every Vulkan application loads the hook even if OBS isn't around to need it. The Vulkan loader iterates over all registered implicit layers, and loads them as part of instance creation (if I recall correctly). More info:

https://renderdoc.org/vulkan-layer-guide.html
https://vulkan.lunarg.com/doc/sdk/1.2.141.2/windows/loader_and_layer_interface.html

Every implicit layer is required to specify an environment variable in its manifest to opt out. Our particular implicit layer can be disabled by specifying DISABLE_VULKAN_OBS_CAPTURE=1 in the environment. I've seen people ask Khronos/LunarG to provide a hammer to opt out of all layers for testing ease, but I don't think such a feature has been implemented yet.

As far as whether the impact of this bug is wide-reaching, it seems unlikely. Most Vulkan applications are games that use one device, and one swap chain, and it's hard to imagine that many applications would push the current 16 limit. In general, we are mindful that being an implicit layer comes with a certain level of responsibility, and try to write robust code.

It would probably be better if the hook didn't have such a generic name. I think something like obs-game-capture-hook64.dll would be more informative. It's not like we're trying to trick anyone.

**Comment by RikoOphorst** [comment]
>What version of Windows was this? Windows 7, 8, 8.1, 10? What version of OBS did you have installed? Was it a release version or a self-compiled version?

For what it's worth, I was on Windows 10, latest version. OBS version I was using was the latest release version as downloadable from the OBS website.

@jpark37 I imagine this issue wasn't due to the sheer number of devices/contexts - I don't think we would exceed that number to be honest. I think the real bit of code that has to be evaluated is the interception that happens around our `vkCreateSwapchainKHR` call - there is likely some logic there that triggers that error log `VkSwapchainKHR was not created, allocated or retrieved from the correct device.` from the validation layer - likely something happens in your layer that doesn't create the swapchain from the correct device or instance.

Upon our `vkCreateSwapchainKHR` , it gets intercepted, and you trigger a `vkGetSwapchainImagesKHR` call (from inside the graphics-hook64.dll layer) that likely is being ran from the wrong `VkDevice` or `VkInstance`, triggering the validation layer error.

Hope that helps :)

**Comment by jpark37** [comment]
It looks like multiple device support has been busted from day 1. I have a modified sample application that reproduces the issue, and it seems to be corrected as a result of the changes in #3150.

**Comment by RikoOphorst** [comment]
@jpark37  Awesome, thanks for the fix!
